File: blind-75/min_stack.java
// leetcode 155

import java.util.*;

class MinStack {
	Stack<Integer> minStack;
	Stack<Integer> stack;

	MinStack () {
		this.stack = new Stack<>();
		this.minStack = new Stack<>();
	}

	public void push (int val) {
		this.stack.add(val);

		if (this.minStack.isEmpty()) {
			this.minStack.add(val);
		} else {
			this.minStack.add(Math.min(this.minStack.peek(), val));
		}
	}

	public void pop () {
		this.stack.pop();
		this.minStack.pop();
	}

	public int top () {
		return this.stack.peek();
	}

	public int getMin () {
		return this.minStack.peek();
	}

}


class Main {
	public static void main (String[] args) {
		MinStack obj = new MinStack();
		
		String[] ops = {"MinStack","push","push","push","getMin","pop","top","getMin"};
		int[][] values = {{}, {-2}, {0}, {-3}, {}, {}, {}, {}};
		List<Integer> res = new ArrayList<>();

		for (int i = 0; i < ops.length; i++) {
			switch(ops[i]) {
				case"push":
					obj.push(values[i][0]);
					res.add(null);
					break;
				case "pop":
					obj.pop();
					res.add(null);
					break;
				case "top":
					res.add(obj.top());
					break;
				case "getMin":
					res.add(obj.getMin());
					break;
			}
		}

		System.out.println(res);
	}
}

File: blind-75/min_stack.py
# leetcode 155

from collections import deque


'''
Approach used maintaining double stacks. One stack normally stores values and another
stores the min value corresponding to each value in the normal stack.
Suppose stack gets -2, so -2 gets into the minStack as min value,
Now stack gets 0, the minStack still gets -2 as the min value. At this point stack is [-2, 0]
And minStack is [-2,-2]
Now stack gets -3, so the current values will be [-2, 0, -3] and [-2, -2, -3] respectively
'''
class MinStack:
	def __init__(self):
		self.stack = deque()
		self.minStack = deque()

	def push(self, val: int) -> None:
		self.stack.append(val)

		if len(self.minStack) == 0:
			self.minStack.append(val)
		else:
			self.minStack.append(min(self.minStack[-1], val))

	def pop(self) -> None:
		self.stack.pop()
		self.minStack.pop()

	def top(self) -> int:
		return self.stack[-1]

	def getMin(self) -> int:
		return self.minStack[-1]


obj = MinStack()

ops = ["MinStack","push","push","push","getMin","pop","top","getMin"]
values = [[],[-2],[0],[-3],[],[],[],[]]

res = list()
for i, op in enumerate(ops):
	if op == "push":
		obj.push(values[i][0])
		res.append(None)
	elif op == "pop":
		obj.pop()
		res.append(None)
	elif op == "top":
		res.append(obj.top())
	elif op == "getMin":
		res.append(obj.getMin())

print(res)